src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,271 @@
+ *   x: number,
+ *   y: number,
+ * }}
+ */
+export let DOMRect;
+
+/**
+ * The IntersectionObserverPolyfill class lets any element receive its
+ * intersection data with the viewport. It acts like native browser supported
+ * IntersectionObserver.
+ * The IntersectionObserver receives a callback function and an optional option
+ * as params. Whenever the element intersection ratio cross a threshold value,
+ * IntersectionObserverPolyfill will call the provided callback function with
+ * the change entry.
+ */
+export class IntersectionObserverPolyfill {
+  /**
+   * @param {!function()} callback.
+   * @param {Object=} opt_option
+   */
+  constructor(callback, opt_option) {
+    /** @private @const {function()} */
+    this.callback_ = callback;
+
+    /**
+     * A list of threshold, sorted in increasing numeric order
+     * @private @const {!Array}
+     */
+    this.threshold_ = opt_option ? opt_option.threshold : [0];
+    this.threshold_ = this.threshold_.sort();
+    dev().assert(this.threshold_[0] >= 0 &&
+        this.threshold_[this.threshold_.length - 1] <= 1,
+        'Threshold should be in the range from "[0, 1]"');
+
+    /** @private {?Element} */
+    this.element_ = null;
+
+    /** @private {boolean} */
+    this.isAmpElement_ = false;
+
+    /**
+     * The prev threshold slot which the previous ratio fills
+     * Range [0, this.threshold_.length]
+     * TODO: Do we always want to call callback at the first tick?
+     * @private {number}
+     */
+    this.prevThresholdSlot_ = 0;
+
+    // TODO: create the PositionObserver class, or create listener for position
+    // info if inside an iframe
+
+    // TODO: Add the postMessageApi logic outside IntersectionObserverPolyfill.
+  }
+
+  /**
+   * Provide a way to observer the intersection change for a specific element
+   * Please note that the IntersectionObserverPolyfill only allow to observe one
+   * element at a time.
+   * TODO: Support observing multiple elements.
+   * TODO: Support non AMP element
+   * @param {!Element} element
+   */
+  observe(element) {
+    // Check the element is an AMP element
+    dev().assert(element.isBuilt && element.isBuilt());
+    // Reset cached value for a new element
+
+    this.prevThresholdSlot_ = 0;
+    this.element_ = element;
+  }
+
+  // TODO: Support unobserve() function
+
+  /**
+   * Tick function that update the DOMRect of the root of observed element.
+   * Caller needs to make sure to pass in the correct container.
+   * Note: the opt_iframe param is the iframe position relative to the host doc,
+   * The iframe must be a non-scrollable iframe.
+   * @param {!./layout-rect.LayoutRectDef} viewport.
+   * @param {./layout-rect.LayoutRectDef=} opt_iframe.
+   */
+  tick(hostViewport, opt_iframe) {
+    if (opt_iframe) {
+      // If element inside an iframe. Adjust origin to the iframe.left/top.
+      hostViewport =
+          moveLayoutRect(hostViewport, -opt_iframe.left, -opt_iframe.top);
+      opt_iframe =
```

Comment:
here's what I thought. It doesn't matter what is the new origin I choose. I can either adjust `opt_iframe` by the `hostViewport` bound or adjust `hostViewport` by the `opt_iframe` bound as long as every rect are calculated later using the same origin. 
In the case of having an `opt_iframe`, `ownerRect` and `elementRect` both have position relative to `opt_iframe` And the returned `boundClientRect` and `intersectionRect` should also have position relative to `opt_iframe` So I assume it would be better to adjust `hostViewport` by `opt_iframe` here instead.